<!doctype html>
<html lang="en">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="description" content="">
    <meta name="author" content="Mark Otto, Jacob Thornton, and Bootstrap contributors">
    <meta name="generator" content="Hugo 0.84.0">
    <title>About PanKB</title>

    <link rel="canonical" href="https://getbootstrap.com/docs/5.0/examples/navbar-fixed/">

    <link href="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-T3c6CoIi6uLrA9TneNEoa7RxnatzjcDSCmG1MXxSR1GAsXEV/Dwwykc2MPK8M2HN" crossorigin="anonymous">

    <script src="https://cdn.jsdelivr.net/npm/bootstrap@5.3.2/dist/js/bootstrap.bundle.min.js" integrity="sha384-C6RzsynM9kWDrMNeT87bh95OGNyZPhcTNXj1NW7RuBCsyN/o0jlpcV8Qyq46cDfL" crossorigin="anonymous"></script>

    <style>

        html {
            font-size: 18px;
        }

        .fixed-sidebar {
            position: fixed;
            top: 100px; /* Adjust if necessary, e.g., below a top navbar */
            left: 0;
            height: 100vh;
            width: 23%;;
            overflow-y: auto; /* To make sidebar scrollable if content is long */
        }

        .navbar-nav .nav-item .nav-link {
            margin-left: 1em;
            margin-right: 1em;
            font-size: 1em;
        }

        .logo-image_navbar {
            height: 3vh;
        }

        h1 {
            text-align: center;
            font-size: 2em; /* Adjust the font-size as needed */
        }

        h2 {
            text-align: center;
            font-size: 1.8em; /* Adjust the font-size as needed */
        }

        h3 {
            text-align: center;
            font-size: 1.6em; /* Adjust the font-size as needed */
        }

    </style>

    <!-- Custom styles for this template -->
    <link href="navbar-top-fixed.css" rel="stylesheet">
</head>

<body data-bs-spy="scroll" data-bs-target="#list-example" data-bs-offset="0">

<nav class="navbar navbar-expand-md navbar-dark fixed-top bg-dark">
    <div class="container-fluid">
        <a class="navbar-brand" href="./index.html" style="margin: 0;">
            <img src="pankb_logo.svg" alt="PanKB" class="d-inline-block align-text-top logo-image_navbar">
        </a>
        <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation">
            <span class="navbar-toggler-icon"></span>
        </button>
        <div class="collapse navbar-collapse justify-content-center" id="navbarCollapse">
            <ul class="navbar-nav mb-2 mb-md-0">
                <li class="nav-item">
                    <a class="nav-link active" aria-current="page" href="./About.html">About</a>
                </li>
                <li class="nav-item">
                    <a class="nav-link " href="./Species_list.html">Organisms</a>
                </li>
<!--                <li class="nav-item">-->
<!--                    <a class="nav-link " href="#">Download</a>-->
<!--                </li>-->
                <li class="nav-item">
                    <a class="nav-link " href="./Publications.html" tabindex="-1" aria-disabled="true">Publications</a>
                </li>
            </ul>
        </div>
        <a class="navbar-brand" href="./index.html" style="margin: 0; visibility: hidden; ">
            <img src="pankb_logo.svg" alt="PanKB" class="d-inline-block align-text-top logo-image_navbar">
        </a>
    </div>
</nav>

<div class="row">
    <div class="col-1">
<!--        <div id="list-example" class="list-group">-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1.1em' href="#about_pankb">About PanKB</a>-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1.1em' href="#how_to_use">How to use PanKB</a>-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1.1em' href="#Pangenome_Analytics_Dashboards">Pangenome Analytics Dashboards</a>-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1em' href="#overview">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Overview</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#heaps_law_and_gene_frequency">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Heap's Law and Gene Frequency</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#presence_absence_matrix">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Presence/Absence Matrix</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#genome_gene_info_page">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genome & Gene Info Page</a>-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1em' href="#genee">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Genes</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#gene_annotation_able">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene Annotation Table</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#gene_info_page">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Gene Info Page</a>-->
<!--            <a class="list-group-item list-group-item-action" style = 'font-size: 1em' href="#phylogenetic_tree">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; Phylogenetic Tree</a>-->
<!--            <a class="list-group-item list-group-item-action" href="#genome_info_page">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Genome Info Page</a>-->
<!--        </div>-->
    </div>

<!--    <div class="col-3 vh-100"></div>-->

    <div class="col-10">
        <div data-bs-spy="scroll" data-bs-target="#list-example" data-bs-offset="0" data-bs-smooth-scroll="true" class="scrollspy-example" tabindex="0">
            <h1 id="about_pankb" style = 'text-align: center'>About PanKB</h1>
            <br>
            <p style = 'font-size: 1.2em'>
                Since the assembly of <a href="https://doi.org/10.1007%2Fs10142-015-0433-4"> the first complete
                microbial genome in 1995</a>,
                the emergence of efficient and inexpensive sequencing technologies has driven the rapid expansion of
                publicly available microbial genome data, with <a
                    href="https://www.biorxiv.org/content/10.1101/2024.03.08.584059v1">almost than 2M microbial genomes
                in public databases</a>.
                The rapid accumulation of microbial genome data presents an opportunity to unlock the vast potential of
                these ubiquitous organisms. The burgeoning field of pangenomics offers a framework for extracting value
                from this big biological dataset. By comparing genomes across strains, pangenomics reveals the genomic
                basis for diverse phenotypes, leading to a deeper understanding of microbial functions. This knowledge
                can be leveraged to drive innovation across multiple application domains, including food production,
                human health, and ecological sustainability, benefiting both humanity and the environment.
            </p>

            <ul style = 'font-size: 1.2em'>
                <li>A growing dataset of pangenomic results,</li>
                <li>Interactive data reports and analytics for exploration, analysis, and new potential discoveries,
                </li>
                <li>Global database search for genes, pathways, products, species, and more,</li>
                <li>Alleleomes describing the amino acid variants across gene alleles,</li>
                <li>Dataset download providing access to raw data, search results, and pangenomics
                    analytics for custom analysis
                </li>
                <li>A bibliome of open-access pangenomic publications,</li>
                <li>A specialized LLM-powered chat interface to accelerate knowledge acquisition by providing
                    deeply-detailed query responses on publication content, including supporting references,
                    and constrained to not return hallucinated content.
                </li>
            </ul>

            <p style = 'font-size: 1.2em'>
                Combined, PanKB's features enable valuable workflows for enzyme and strain engineering. These include
                identifying genes for new enzyme production or reintroduction into strains, pinpointing precise gene
                edits to modify activity, discovering and optimizing valuable pathways, and selecting optimal starting
                strains. Current strain engineering heavily relies on model or familiar strains; the features and data
                of PanKB empower strain engineers to start leveraging pangenomic data for targeted bioengineering.
            </p>

            <br>

            <h1 id="how_to_use" style = 'text-align: center'>Using PanKB</h1>
            <br>
            <h3 style='text-align: center'>Pangenomes</h3>
            <p style = 'font-size: 1.2em'>
                PanKB includes multiple interactive analytics and tables that give an overview of a pangenome's
                contents. These analytics can be found or navigated to on an individual pangenome's page.
            </p>

            <h3 style='text-align: center'>Alleleomes</h3>
            <p style = 'font-size: 1.2em'>
                Pangenomes also serve as the foundation for further large-scale analyses, and PanKB is actively
                integrating their novel results. Recent pangenomic-scale analyses of variants, named <a
                    href="https://doi.org/10.1073/pnas.2218835120">Alleleomics</a>,
                demonstrated unique value in <a href="https://www.biorxiv.org/content/10.1101/2023.09.17.558058v1">narrowing
                the solution search space for feasible genetic variants in <i>E.coli</i> </a>. PanKB currently includes the alleleomes of all of its pangenomes. Alleleome analytics
                can be found on pangenome and specific gene pages.
            </p>

            <h3 style='text-align: center'>Data Accessibility</h3>
            <p style = 'font-size: 1.2em'>
                PanKB implements multiple different methods to accessing its data. Users can access all of PanKB's data
                through its navigation links. Users can also quickly find specific data through the global database
                search
                feature accessible on most pages. Finally, users can also download the database's raw data
                through the various analytics hosted on database pages.
            </p>

            <h3 style='text-align: center'>PanKB LLM</h3>
            <p style = 'font-size: 1.2em'>
                Scientific progress often necessitates extensive literature review, a traditionally time-consuming
                process. Large Language Models (LLMs) offer a potential solution by aggregating and summarizing
                knowledge across documents. PanKB includes an LLM chatbot (AI Assistant) focused on an open-access
                pangenomic bibliome,
                designed to accurately answer deep questions on pangenomics, cite relevant articles, and not attempt to
                hallucinate inaccurate content. This feature is an initial experiment towards combining an LLM and a
                specialized scientific database to accelerate scientific knowledge acquisition through automated
                knowledge extraction.
            </p>

            <br>

            <h2 style='text-align: center'>Tools and Methods</h2>
            <p style = 'font-size: 1.2em'>
                 <a href="https://doi.org/10.1093/nar/gkae314">BGCFlow</a>: Nuhamunada M, Mohite OS, Phaneuf PV, Palsson BO,
                Weber T. BGCFlow: systematic pangenome workflow for the analysis of biosynthetic gene clusters across
                large
                genomic datasets. Nucleic Acids Research. 2024 Jun 10;52(10):5478–95.
            </p>
            <h3 style='text-align: center'>Collected Pangenomic Publication</h3>
            <p style = 'font-size: 1.2em'>
                An interactive table for these can be found in the <a href=" /publications/ ">Collected Publications</a> page.
            </p>
            <br>
            <h2 style='text-align: center'>Contact Us</h2>
            <p style = 'font-size: 1.2em'>
                If you have questions or find any bugs in the database, please contact <a href="https://orbit.dtu.dk/en/persons/patrick-victor-phaneuf">Patrick Phaneuf</a>

            </p>
            <br>

<!--            <p style = 'font-size: 1.2em'>-->
<!--                To begin, choose an organism of interest from the organism list. You can access the organism list page by either clicking the 'Select an organism' button located at the bottom of the landing page or selecting the 'Organism' option in the navbar. Next, you can pick an organism of interest and explore its pangenomic analytics by clicking on the link in the last column labeled 'Pangenome analyses.' This action will direct you to the Pangenome Analysis Dashboards.-->
<!--                <br>-->
<!--            <div style="text-align: center;">-->
<!--                <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/pankb_workflow.png" style="width: 80%;">-->
<!--            <br>-->
<!--                <p id="imageFootprint"  style = 'font-size: 1.3em' >PanKB workflow</p>-->
<!--            </div>-->

<!--                <br>-->
<!--                <br>-->
<!--                Alternatively, you can access the overview page of a specific species by clicking on the data point within the 'Pangenomes' figure located in the middle of the landing page.-->
<!--            </p>-->

<!--            <br>-->

<!--            <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Landing_page.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--            <br>-->

<!--            <h1 id="Pangenome_Analytics_Dashboards" style = 'text-align: center'>Pangenome Analytics Dashboards</h1>-->

<!--            <br>-->

<!--            <p style = 'font-size: 1.3em'>-->
<!--                Each dashboard showcases the pangenome analysis results for a organism species. It consists of three parts: Overview, Genes, and Phylogenetic Tree. In the following section, we provide examples of each page element from the <a href = 'https://pankb.org/Pangenome_analyses_overview.html?species=Lactiplantibacillus_plantarum'>Lactiplantibacillus plantarum Pangenome Analytics Dashboards</a>.-->
<!--                <br>-->
<!--                <br>-->
<!--                By default, when you enter an organism's Pangenome Analytics Dashboards, you will first see the overview page. You can navigate to the other two pages using the sidebar on the left. </p>-->

<!--            <h2 id="overview" style = 'text-align: center;' >Overview</h2>-->

<!--            <br>-->

<!--            <p  style = 'font-size: 1em'> The <b>Overview</b> page contains several interactive figures that depict various pangenome properties of an organism. </p>-->

<!--            <h3 id="heaps_law_and_gene_frequency" style = 'text-align: center;' >Heap's Law and Gene Frequency</h3>-->

<!--            <br>-->

<!--            <p  style = 'font-size: 1em'>-->

<!--                The first 4 figures in the <b>Overview</b> page display the four standard calculations related to elementary pangenome properties: Heap's Law, Cumulative Gene Distribution, Gene Annotation Distribution, and Gene Frequency. All of them are interactive figures that offer the following functionalities:-->
<!--                <br>-->
<!--                <b>1. Hovering:</b> By using the mouse to hover over a figure element (line, bar, etc.), you can access more details about that particular element.-->
<!--                <br>-->
<!--                <b>2. Hide and show:</b> Clicking on the legends allows you to hide or show the corresponding Gene class (Core, Accessory, Rare) in the figure.-->
<!--                <br>-->
<!--                <b>3.Zoom in and Zoom out:</b> If you wish to zoom in on a specific area of a figure, simply press and hold the left mouse button within the figure while selecting the corresponding area. To undo the zoom in, just double-click the left mouse button.-->
<!--            </p>-->

<!--            <br>-->

<!--                <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Heaps_law_gene_frequency.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--            <br>-->

<!--            <h3 id="presence_absence_matrix" style = 'text-align: center;' >Presence/Absence Matrix</h3>-->
<!--            <br>-->
<!--            <p  style = 'font-size: 1em'>-->
<!--                The Presence/Absence Matrix is represented as an interactive heatmap. By default, the heatmap displays the Presence/Absence Matrix only for Core genes. To explore other gene classes such as Accessory, Rare, etc., you can use the <b>'Select a Gene Class'</b> button located in the upper left corner of the figure.-->
<!--                <br>-->
<!--                <br>-->
<!--                As an interactive heatmap, the Presence/Absence matrix has the following features:-->
<!--                <br>-->
<!--                <b>1. Zooming In and Out:</b> You can zoom in and out on the x-axis and y-axis of the heatmap by dragging the two sliders located in the upper left corner of the heatmap. The horizontal and vertical lines surrounding the small lock icon correspond to these axes. To simultaneously zoom in on both the x-axis and y-axis, simply click on the small lock icon and then drag either of the two lines.-->
<!--                <br>-->
<!--                <b>2. Hover Functionality:</b> Hover your mouse over a cell in the heatmap to view information related to the corresponding gene (x-axis label) and genome (y-axis label).-->
<!--                <br>-->
<!--                <b>3. Links to Other Pages:</b> Click on a cell of interest to navigate to the genome or gene page that corresponds to that cell. On these pages, you can access more detailed information about the genome and gene (refer to the following section: Genome and Gene Info Page).-->
<!--            </p>-->

<!--            <br>-->

<!--            <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Heatmap.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--            <br>-->

<!--            <h3 id="genome_gene_info_page" style = 'text-align: center;' >Genome & Gene Info Page</h3>-->
<!--            <br>-->
<!--            <p  style = 'font-size: 1em'>-->
<!--                This page offers more information about the genome and genes from the Presence/Absence Matrix. The page mainly consists of three parts:-->
<!--                <br>-->
<!--                <b>1. Genome and Gene Tables:</b> These two tables contain basic information about the genome and genes. The genome table includes a link to NCBI, where you can find more information about this genome. The gene table also provides a link to the Gene page (See <a href = '#gene_info_page'>Gene Info Page</a>).-->
<!--                <br>-->
<!--                <b>2. Gene Annotation Distribution:</b> This is a similar interactive bar plot to the Gene Annotation Distribution on the Overview page. However, this plot focuses on a single genome, whereas the one on the Overview page covers the entire species.-->
<!--                <br>-->
<!--                <b>3. Secondary Metabolite BGCs:</b> This section displays the results generated by antiSMASH. If you would like to learn more about antiSMASH, please visit <a href = 'https://antismash.secondarymetabolites.org/#!/about'>antiSMASH</a>.-->

<!--            <br>-->
<!--            <br>-->
<!--            <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Genome_gene_info.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--            <br>-->

<!--            </p>-->
<!--            <h2 id="genes" style = 'text-align: center'>Genes</h2>-->
<!--            <p style = 'font-size: 1em'> <b>Genes</b> page mainly show the annotation of genes. </p>-->

<!--            <h3 id="gene_annotation_able" style = 'text-align: center;' >Gene Annotation Table</h3>-->
<!--            <br>-->
<!--            <p  style = 'font-size: 1em'>-->
<!--                The Gene Annotation table is an interactive table on the Genes page. It possesses the following features:-->
<!--                <br>-->
<!--                <b>1.Sorting:</b> You can sort the entire table based on any single column.-->
<!--                <br>-->
<!--                <b>2.Searching:</b> You can enter filter conditions in the search box of each column to precisely filter the data you desire.-->
<!--                <br>-->
<!--                <b>3.Links to other pages:</b> Each entry in the first column of the Gene Annotation Table serves not only as the gene name but also as a link to the corresponding <a href = '#gene_page'>Gene Info Page</a>. These gene pages contain more comprehensive gene information and advanced pangenome analyses.-->
<!--                <br>-->
<!--                <br>-->
<!--                <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Gene_annotation.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--                <br>-->

<!--            <h3 id="gene_info_page" style = 'text-align: center;' >Gene Info Page</h3>-->
<!--            <br>-->
<!--            <p  style = 'font-size: 1em'>-->
<!--                The Gene page currently comprises two components:-->
<!--                <br>-->
<!--                <b>1. Gene Information Table:</b> This interactive table contains all Locustags related to this gene, along with their corresponding Genome ID (found in the second column labeled 'Genome ID' which also serves as a link to the respective Genome Info Page), annotation details, and sequence information.-->
<!--                <br>-->
<!--                <b>2. Multiple Sequence Alignment:</b> Each core gene is accompanied by a Multiple Sequence Alignment. This alignment is presented as an interactive figure with the following features:-->
<!--                <br>-->
<!--                &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<b>1.Overview Box:</b> Located at the bottom of the figure, the overview box allows you to navigate the multiple sequence alignment plot. You can also select a specific area on the overview box using the left mouse button, and the multiple sequence alignment will jump to the selected region.-->
<!--                <br>-->
<!--                &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<b>2.Scrolling Function:</b> You can scroll up, down, left, and right by holding down the left mouse button and dragging the image.-->
<!--                <br>-->
<!--                &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<b>3.Amino Acid Percentage:</b> Clicking on the position number (found below the logo plot) enables you to view the amino acid percentage for each position.-->
<!--                <br>-->
<!--                &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<b>4.Links to Other Pages:</b> You can access the corresponding Genome's Genome Info Page by clicking on the Genome ID along the y-axis. The Genome Info Page is similar to the Genome and Gene Information Page but does not include the gene table (refer to the above section for the link).-->
<!--                <br>-->
<!--                <br>-->
<!--                <img src="https://pankb.blob.core.windows.net/data/PanKB/web_data/about_page_gif/Gene_info.gif" style="display: block; margin-left: auto; margin-right: auto; width: 80%;">-->

<!--                <br>-->
<!--            </p>-->

<!--            <h2 id="phylogenetic_tree" style = 'text-align: center'>Phylogenetic Tree</h2>-->
<!--            <p style = 'font-size: 1em'> The Phylogenetic Tree page displays the phylogenetic traits of a species, and it is constructed using the Mash distance method. The phylogenetic tree in PanKB is generated using Phylotree.js. You can find additional information about its features in <a href = 'https://github.com/veg/phylotree.js/'>phylotree.js</a>. In addition to the features documented in the Phylotree.js documentation, we have also incorporated links to the Genome Info Page. By left-clicking on the genome of interest and selecting 'Search for this strain,' you can navigate to the corresponding Genome Info Page and access more information about that specific genome. </p>-->


<!--            <h3 id="genome_info_page" style = 'text-align: center;' >Genome Info Page</h3>-->
<!--            <br>-->
<!--            <p  style = 'font-size: 1em'>-->
<!--                The Genome Info Page is similar to the <a href = '#genome_gene_info_page'>Genome & Gene Info Page</a>, just without gene table.-->
<!--            </p>-->

        </div>
    </div>

    <div class="col-1 vh-100"></div>
</div>

<script>
    document.addEventListener("DOMContentLoaded", function(){
        const scrollSpyEl = document.body; // or document.querySelector('.scrollspy-example')
        const scrollSpy = new bootstrap.ScrollSpy(scrollSpyEl, {
            target: '#list-example'
        });
    });
</script>

<script>
    function adjustHtmlFontSize(coefficient) {
        const newFontSize = 18 * coefficient;
        document.documentElement.style.fontSize = newFontSize + 'px';
    }

    // Function to adjust button styles
    function adjustButtonStyles(coefficient) {
        const buttons = document.getElementsByClassName('link_button');
        for (let i = 0; i < buttons.length; i++) {
            buttons[i].style.padding = (18 * coefficient) + 'px ' + (18 * coefficient) + 'px';
            buttons[i].style.fontSize = (20 * coefficient) + 'px';
        }
    }


    function adjustStyles() {
        const template_screen_width = 1920;
        const current_window_Width = window.innerWidth;
        const coefficient = current_window_Width / template_screen_width;

        adjustHtmlFontSize(coefficient);
        adjustTitleFontSize(coefficient);
        adjustButtonStyles(coefficient);
    }

    document.addEventListener("DOMContentLoaded", adjustStyles);
    window.addEventListener("resize", adjustStyles);
</script>



</body>
</html>
